contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,202 @@
+        require(newBufferCap <= bufferCap, "MultiRateLimited: new buffer cap is over global max");
+
+        _setIndividualBufferCap(rateLimitedAddress, newBufferCap);
+    }
+
+    /// @notice the amount of action used before hitting limit
+    /// @dev replenishes at rateLimitPerSecond per second up to bufferCap
+    /// @param rateLimitedAddress the address whose buffer will be returned
+    function individualBuffer(address rateLimitedAddress) public view override returns(uint144) {
+        RateLimitData memory rateLimitData = rateLimitPerAddress[rateLimitedAddress];
+
+        uint256 elapsed = block.timestamp - rateLimitData.lastBufferUsedTime;
+        return uint144(Math.min(rateLimitData.bufferStored + (rateLimitData.rateLimitPerSecond * elapsed), rateLimitData.bufferCap));
+    }
+
+    /// @notice the rate per second for each address
+    function individualRateLimitPerSecond(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].rateLimitPerSecond;
+    }
+
+    /// @notice the last time the buffer was used by each address
+    function individualLastBufferUsedTime(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].lastBufferUsedTime;
+    }
+
+    /// @notice the cap of the buffer that can be used at once
+    function individualBufferCap(address limiter) external override view returns(uint256) {
+        return rateLimitPerAddress[limiter].bufferCap;
+    }
+
+    /// @notice the method that enforces the rate limit. Decreases buffer by "amount". 
+    /// If buffer is <= amount either
+    /// 1. Does a partial depletion by the amount remaining in the buffer or
+    /// 2. Reverts
+    /// Depending on whether doPartialAction is true or false
+    /// @param rateLimitedAddress the address whose buffer will be depleted
+    /// @param amount the amount to remove from the rateLimitedAddress's buffer
+    function _depleteBuffer(address rateLimitedAddress, uint256 amount) internal returns(uint256) {
+        _depleteBuffer(amount);
+
+        uint256 newBuffer = individualBuffer(rateLimitedAddress);
+        
+        uint256 usedAmount = amount;
+        if (doPartialAction && usedAmount > newBuffer) {
+            usedAmount = newBuffer;
+        }
+
+        require(newBuffer != 0, "MultiRateLimited: no rate limit buffer");
+        require(usedAmount <= newBuffer, "MultiRateLimited: rate limit hit");
+
+        rateLimitPerAddress[rateLimitedAddress].bufferStored = uint144(newBuffer - usedAmount);
+
+        rateLimitPerAddress[rateLimitedAddress].lastBufferUsedTime = _blockTimestamp();
+
+        emit IndividualBufferUsed(rateLimitedAddress, usedAmount, newBuffer - usedAmount);
+
+        return usedAmount;
+    }
+
+    /// @notice return current block timestamp as a uint32
+    function _blockTimestamp() private view returns (uint32) {
```

Comment:
This function is one line, I don't see a need for it

Reply:
its cleaner to call `_blockTimestamp()` instead of this `block.timestamp.toUint32()`

Reply:
Disagree tbh. But it also adds gas as it's an extra function call (unless the compiler embeds this inline, which I don't think it does)

Reply:
since when have we been concerned about optimizing down to the last opcode? Our code isn't user facing so it's fine

Reply:
You're positing a position I'm not taking; this is an easy and simple optimization. We should take these when we see them.